# Linear Regression from Scratch using Gradient Descent

In [1]:
#Import the required libraries
from numpy import *

#### Step 1: Collect Data

In [2]:
points=genfromtxt('Points.txt', delimiter=',')

#### Step 2: Define the hyperparameters

In [3]:
learning_rate=0.0001            #How should our model converge
initial_b=0
initial_m=0
num_iterations=1000

#### Step 3: Train our model

In [4]:
#Compute the error using sum of squared differences
def compute_error(b,m,points):
    total_error=0
    for i in range(len(points)):
        y=points[i,1]
        x=points[i,0]
        
        #calculate the error
        total_error+=(y - (m * x + b)) ** 2
        
        #calculate the average
    return total_error/float(len(points))

#Calculate the gradient descent
def gradient_descent(points, starting_b, starting_m, learning_rate, num_iterations):
    #initialize b and m
    b=starting_b
    m=starting_m
    
    #gradient descent
    for i in range(num_iterations):
        #update b and m with more accurate value
        b, m = gradient_step(b, m, points, learning_rate)
    return [b,m]

def gradient_step(b_current, m_current, points, learningRate):
    
    #starting points of gradients
    b_gradient=0
    m_gradient=0
    n=float(len(points))
    
    for i in range(0, len(points)):
        x= points[i,0]
        y= points[i,1]
        
        #calculating derivatives for finding minimum cost function
        b_gradient+=-(2/n) * (y - ((m_current*x) + b_current))
        m_gradient+=-(2/n) * x * (y - ((m_current*x) + b_current))
        
    #update b and m values using partial derivatives
    new_b=b_current - (learningRate*b_gradient)
    new_m=m_current - (learningRate*m_gradient)
    return [new_b, new_m]

#### Step 4: Execute the model

In [5]:
print('Starting gradient descent at b={0}, m={1}, error={2}'.format(initial_b,initial_m, compute_error(initial_b, initial_m, points)))

[b,m] = gradient_descent(points, initial_b, initial_m, learning_rate, num_iterations)

print('Running...')

print('After {0} iterations, b={1}, m={2}, error={3}'.format(num_iterations,b,m, compute_error(b, m, points)))

Starting gradient descent at b=0, m=0, error=5143.248003616552
Running...
After 1000 iterations, b=0.03874859049412073, m=1.3713203229029995, error=453.7412322181122


Thus, with the help of gradient descent, we reduced the error from 5143.24 to 453.74.